File: stack/csharp/ReverseSomeChars.cs
using System.Collections.Generic;

namespace StackSolutions;

public static class ReverseSomeChars
{
    public static string Reverse(string s, IEnumerable<char> chars)
    {
        var set = new HashSet<char>(chars);
        var stack = new Stack<char>();
        foreach (var ch in s)
        {
            if (set.Contains(ch))
            {
                stack.Push(ch);
            }
        }

        var result = new char[s.Length];
        for (var i = 0; i < s.Length; i++)
        {
            var ch = s[i];
            result[i] = set.Contains(ch) ? stack.Pop() : ch;
        }

        return new string(result);
    }
}

File: stack/java/ReverseSomeChars.java
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Set;

public final class ReverseSomeChars {
  private ReverseSomeChars() {
  }

  public static String reverse(String s, Set<Character> chars) {
    Deque<Character> stack = new ArrayDeque<>();
    for (char ch : s.toCharArray()) {
      if (chars.contains(ch)) {
        stack.push(ch);
      }
    }

    StringBuilder result = new StringBuilder(s.length());
    for (char ch : s.toCharArray()) {
      if (chars.contains(ch)) {
        result.append(stack.pop());
      } else {
        result.append(ch);
      }
    }

    return result.toString();
  }
}
